Detectors/TPC/reconstruction/src/HwClusterer.cxx

```diff
@@ -268,108 +253,157 @@ void HwClusterer::processDigits(
 //________________________________________________________________________
-void HwClusterer::Process(std::vector<o2::TPC::Digit> const &digits)
+void HwClusterer::Process(std::vector<o2::TPC::Digit> const &digits, MCLabelContainer const* mcDigitTruth, int eventCount)
 {
   mClusterArray->clear();
+  if(mClusterMcLabelArray) mClusterMcLabelArray->clear();
 
-  /*  
+
+  /*
    * clear old storages
    */
   for (auto& cs : mClusterStorage) cs.clear();
+  for (auto& cdis : mClusterDigitIndexStorage) cdis.clear();
   for (auto& dc : mDigitContainer ) {
     for (auto& dcc : dc) dcc.clear();
   }
 
-//  int iCRU;
-//  int iRow;
-//  int iPad;
-//  float charge;
   int iTimeBin;
   int iTimeBinMin = (mIsContinuousReadout)?mLastTimebin + 1 : 0;
   //int iTimeBinMin = mLastTimebin + 1;
   int iTimeBinMax = mLastTimebin;
 
-  /*  
+  if (mLastMcDigitTruth.find(eventCount) == mLastMcDigitTruth.end())
+    mLastMcDigitTruth[eventCount] = std::make_unique<MCLabelContainer>();
+  gsl::span<const o2::MCCompLabel> mcArray;
+  /*
    * Loop over digits
    */
+  int digitIndex = 0;
   for (const auto& digit : digits) {
+
     /*
      * add current digit to storage
      */
-//    iCRU     = digit->getCRU();
-//    iRow     = digit->getRow();
-//    iPad     = digit->getPad();
-//    charge   = digit->getChargeFloat();
-//    if ((cru == 179)) {// && iRow == 5)){
-//      printf("hw:  digi: %d, %d, %d, %d, %.2f\n", cru, iRow, iPad, iTimeBin, charge);
-//    }
 
     iTimeBin = digit.getTimeStamp();
-    if (iTimeBin < iTimeBinMin) continue;
+    if (digit.getCRU() < static_cast<int>(mCRUMin) || digit.getCRU() > static_cast<int>(mCRUMax)) {
+      LOG(DEBUG) << "Digit [" << digitIndex << "] is out of CRU range (" << digit.getCRU() << " < " << mCRUMin << " or > " << mCRUMax << ")" << FairLogger::endl;
+      // Necessary because MCTruthContainer requires continuous indexing
+      if (mcDigitTruth != nullptr) mLastMcDigitTruth[eventCount]->addElement(digitIndex,o2::MCCompLabel(-1,-1,-1));
+      ++digitIndex;
+      continue;
+    }
+    if (iTimeBin < iTimeBinMin) {
+      LOG(DEBUG) << "Digit [" << digitIndex << "] time stamp too small (" << iTimeBin << " < " << iTimeBinMin << ")" << FairLogger::endl;
+      // Necessary because MCTruthContainer requires continuous indexing
+      if (mcDigitTruth != nullptr) mLastMcDigitTruth[eventCount]->addElement(digitIndex,o2::MCCompLabel(-1,-1,-1));
+      ++digitIndex;
+      continue;
+    }
+
     iTimeBinMax = std::max(iTimeBinMax,iTimeBin);
-    mDigitContainer[digit.getCRU()][digit.getRow()].push_back(&digit);
+    if (mcDigitTruth == nullptr)
+      mDigitContainer[digit.getCRU()][digit.getRow()].emplace_back(std::make_tuple(&digit,-1,eventCount));
+    else {
+      mDigitContainer[digit.getCRU()][digit.getRow()].emplace_back(std::make_tuple(&digit,digitIndex,eventCount));
+      mcArray = mcDigitTruth->getLabels(digitIndex);
+      for (auto &l : mcArray) mLastMcDigitTruth[eventCount]->addElement(digitIndex,l);
```

Comment:
@sawenzel can we have a copy-constructor for the MCTruthContainer? Currently I have to copy it here element by element.

Reply:
@KlewinS : will take a look